<!doctype html>
<html>
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <title>CheckM by Ecogenomics</title>

    <link rel="stylesheet" href="stylesheets/styles.css">
    <link rel="stylesheet" href="stylesheets/pygment_trac.css">
    <meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">
    <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
  </head>
  <body>
    <div class="wrapper">
      <header>
        <h1>CheckM</h1>
        <p>Beta check your meta before you wreck your meta</p>

        <p class="view"><a href="https://github.com/Ecogenomics/CheckM">View the Project on GitHub <small>Ecogenomics/CheckM</small></a></p>


        <ul>
          <li><a href="https://github.com/Ecogenomics/CheckM/zipball/master">Download <strong>ZIP File</strong></a></li>
          <li><a href="https://github.com/Ecogenomics/CheckM/tarball/master">Download <strong>TAR Ball</strong></a></li>
          <li><a href="https://github.com/Ecogenomics/CheckM">View On <strong>GitHub</strong></a></li>
        </ul>
      </header>
      <section>
        <h1>Overview</h1>

<p>One of the problems faced in metagenomic assembly is the uncertainty
that collections of contigs that have been 'binned' into putative
genomes may be contaminated from another organism or represent an
incomplete genome.  Both are major problems that need to be addressed
before inferences about metabolic function can be made by the presence
or absence of genes. CheckM, is a set of tools that can assess the 
contamination and completeness of genome bins using conserved genes found
in a single copy for different taxonomic lineages.  More closely 
related bacteria and archaea share a higher proportion of their gene
complement and therefore estimations of completeness based on these
markers will be more accurate than estimates based on gene markers
that are universal. </p>

<h1>Installation and Dependancies</h1>

<p>CheckM requires both <code>biopython</code> and <code>simpleHMMER</code> python packages be
intalled as well as <code>hmmer3</code> binaries to be in your <code>PATH</code>.  With those
dependancies filling download the git repository to your system and run</p>

<pre><code>python setup.py install
</code></pre>

<h1>Running Instructions</h1>

<p>CheckM has a single executable <code>checkM</code> that has a number of
subcommands.  There are three subcommands that the use will use
regularly:</p>

<ul>
<li>
<code>build</code> - Run <code>hmmer</code> of a marker set against genome(s) of interest</li>
<li>
<code>qa</code> - Filter <code>hmmer</code> results and output a summary</li>
<li>
<code>align</code> - Align identified markers against the HMM model
CheckM can be run in two different ways.  First is for the user to
supply a <code>hmmer3</code> compatible model file of markers.  The second way is
to build a database of markers for all taxonomic lineages and use those
markers for accurate estimates of completeness and contamination.  By
default CheckM prefers to use the later of the two options.</li>
</ul><h2>Setting up the Database</h2>

<p><code>checkM</code> has a subcommand called <code>makeDB</code> that will take freely
available genomic data and calculate the conserved markers for each
taxonomic lineage.  To generate the database you will need to download
the genomes from IMG and have the Pfam-A domains as well.</p>

<p>Assuming that you have either made a database you should set the path
with the environmental variable <code>CHECKM_DB</code>.  To check whether this is
already set, try the following command:</p>

<pre><code>$ echo $CHECKM_DB
</code></pre>

<p>If you see a file path then the CheckM database is already set up on
your system.  Otherwise you can setup the database using the following:</p>

<pre><code>export CHECKM_DB=/full/path/to/your/checkm_database
</code></pre>

<p>To make that setting perminant add the above line into your <code>.bashrc</code> file.</p>

<h1>Marker Sets</h1>

<p>Marker sets are a list of genes/protein domains that are found in all 
genomes for a particular taxonomic rank.  Some marker sets may be 
universal or be specific to a family or genus.  There are two marker
sets that get used routinely; one from
<a href="http://phylosift.wordpress.com">phylosift</a> that contains 37 markers
that are universal, and a list compiled by <a href="http://www.ncbi.nlm.nih.gov/pubmed/?term=22170421">Dupont et al.
(2012)</a> which
contains 111 markers found in the majority of Bacteria. A more accurate
view of contamination and completeness of a genome bin may be achieved
by using markers that are targeted to a particular microbial lineage.</p>

<h2>Generation of taxon specific marker sets</h2>

<p>The taxon specific marker sets that are used by CheckM were generated
using the metadata from the IMG 4.0 release.  First all of the complete
genomes were separated based on the taxonomy string found in the
metadata.  Next the prevalence of pfam domains for each taxon with at
least 5 genomes was calculated todetermine which were found in at least
95% of all genomes.  This list of domainswas further segregated based on
whether that pfam was strictly single copy in each genome or not.</p>

<h3>Caveats</h3>

<ul>
<li>These lists use pfams, which means that a marker may not represent a whole
gene.  For example there may be two markers, one from the C-terminal and
the other from the N-terminal domain of the same protein. </li>
<li>using a minimum of 5 genomes for a taxon is probably too low.  Be careful
about how specific a marker set you use.  Think about the total number of
genomes that the list was constructed against, if in doubt use a Domain
or phylum level marker set.</li>
<li>based solely on the genomes found in IMG 4.0</li>
<li>I haven't done any annotations myself, I'm relying solely on the annotation
pipeline in IMG</li>
<li>The taxonomy is also straight from IMG and this may be inaccurate (maybe?),
which would throw off all of the calculations </li>
</ul><h2>Using Taxon Specific Marker Sets</h2>

<p>Taxon specific marker sets can be invoked by using the <code>-T</code> option on
the command-line.  The taxonomy must be the full taxonomy in Greengenes
format, for example: <code>k__Bacteria;p__Proteobacteria</code>.  Spaces are also
allowed between taxonomic ranks: <code>k__Bacteria; p__Proteobacteria</code>. To
get a universal marker set (common to all bacteria and archaea) use the
special taxonomy 'universal' as the argument to <code>-T</code>
NOTE: Do <strong>not</strong> specify a taxonomy that is lower than the genus level!</p>

<h1>Workflow</h1>

<p>This is a workflow that I use for determining the completeness and
contamination of all of my genome bins.  First start by using one of the
universal marker sets such as the markers used by phylosift or by
specifying the taxonomy as 'universal' in the <code>-T</code> flag of `checkM.</p>

<pre><code>checkM build -T universal &lt;OUTPUT_FOLDER&gt; genome.fna 

checkM qa -T universal &lt;OUTPUT&gt; 
</code></pre>

<p>From the output of the <code>qa</code> command you should be able to see which bins
are 'near complete'.  As a general rule I only look genome bins that are
at least 80% complete and less than 10% contamination.  From this point
I make a tree using the concatenation of the  markers and determine the
putative taxonomy of my genome bins. </p>

<p>For example lets assume that one of your genome bins is a
<em>Rhodocyclacae</em> based on the taxonomy in the tree.  Now we can use the
specific single copy genes for <em>Rhodocyclacae</em> to get more accurate
statistics on completeness and contamination. (There are 428
<em>Rhodocyclacae</em> specific single copy genes, verses the 39 universal)</p>

<pre><code>checkM build -T 'k__Bacteria;p__Proteobacteria;c__Betaproteobacteria;o__Rhodocyclales;f__Rhodocyclaceae' &lt;OUTPUT_FOLDER&gt; genome.fna

checkM qa -T 'k__Bacteria;p__Proteobacteria;c__Betaproteobacteria;o__Rhodocyclales;f__Rhodocyclaceae' &lt;OUTPUT_FOLDER&gt; 
</code></pre>

<p>This output should be a more accurate view of the contamination and completeness of your genome.</p>

<h1>Manual</h1>

<h2>Build</h2>

<pre><code>checkM build [-h] [-T TAXONOMY] [-d DATABASE] [-H HMM] [-b BIN_FOLDER]
                    [-x EXTENSION] [-t THREADS] [-p PREFIX] [-v]
                    out_folder [FASTA [FASTA ...]]

Parse binned contigs

positional arguments:
  out_folder            a place to write output files
  FASTA                 A list of fasta formatted files to be used as the
                        input (default: None)

optional arguments:
  -h, --help            show this help message and exit
  -T TAXONOMY, --tax-string TAXONOMY
                        Specify a Greengenes taxonomy string for the lineage
                        of interest (default: None)
  -d DATABASE, --database DATABASE
                        Specify a different path for the marker database. The
                        default is to use the value of the environmental
                        variable "CHECKM_DB" (default: None)
  -H HMM, --hmm HMM     hmm used in the search (default: None)
  -b BIN_FOLDER, --bin_folder BIN_FOLDER
                        folder containing bins to check (fasta formatted)
                        (default: None)
  -x EXTENSION, --extension EXTENSION
                        used in conjuntion with -b to specify the extension
                        used on files to be parsed (default: fa)
  -t THREADS, --threads THREADS
                        max number of active threads (default: 1)
  -p PREFIX, --prefix PREFIX
                        prefix used for naming output files (default: )
  -v, --verbose         print more (default: False)
</code></pre>

<h2>QA</h2>

<pre><code>checkM qa [-h] [-o OUT_FORMAT] [-T TAXONOMY] [-d DATABASE] [-H HMM]
                 [-p PREFIX] [-e E_VALUE] [-l LENGTH] [-f FILE] [-v]
                 out_folder

Do QA on pre-processed contigs

positional arguments:
  out_folder            folder specified during build command

optional arguments:
  -h, --help            show this help message and exit
  -o OUT_FORMAT, --output-format OUT_FORMAT
                        Change the output format to one of the following: 
                        1. Simple Summary showing bin name, counts, completeness
                           and contamination 
                        2. List of marker names and their counts
                        3. Matrix of marker counts 
                        4. Tabular list of bin name, marker name, contig name
                        5. Tabular list showing contigs that contain more than
                           one copy of the same marker. Format: bin name,
                           contig name, {Marker name, count}... (default: 1)
  -T TAXONOMY, --tax-string TAXONOMY
                        Specify a Greengenes taxonomy string for the lineage
                        of interest (default: None)
  -d DATABASE, --database DATABASE
                        Specify a different path for the marker database. The
                        default is to use the value of the environmental
                        variable "CHECKM_DB" (default: None)
  -H HMM, --hmm HMM     hmm used in the search (default: None)
  -p PREFIX, --prefix PREFIX
                        prefix which was used for naming output files
                        (default: )
  -e E_VALUE, --e_value E_VALUE
                        e value cut off (default: 1e-10)
  -l LENGTH, --length LENGTH
                        percent overlap between target and query (default:
                        0.7)
  -f FILE, --file FILE  print results to file (default: STDOUT)
  -v, --verbose         print more (default: False)
</code></pre>

<h3>Output Format Examples</h3>

<h4>One</h4>

<h4>Two</h4>

<h4>Three</h4>

<h4>Four</h4>

<h4>Five</h4>

<h2>Align</h2>

<pre><code>checkM align [-h] [-T TAXONOMY] [-d DATABASE] [-H HMM] [-p PREFIX]
                    [-e E_VALUE] [-l LENGTH] [-v] [-s] [-o OUT_FORMAT] [-c]
                    [-b]
                    out_folder

Create alignments of hmms to the matched markers

positional arguments:
  out_folder            folder specified during build command

optional arguments:
  -h, --help            show this help message and exit
  -T TAXONOMY, --tax-string TAXONOMY
                        Specify a Greengenes taxonomy string for the lineage
                        of interest (default: None)
  -d DATABASE, --database DATABASE
                        Specify a different path for the marker database. The
                        default is to use the value of the environmental
                        variable "CHECKM_DB" (default: None)
  -H HMM, --hmm HMM     hmm used in the search (default: None)
  -p PREFIX, --prefix PREFIX
                        prefix which was used for naming output files
                        (default: )
  -e E_VALUE, --e_value E_VALUE
                        e value cut off (default: 1e-10)
  -l LENGTH, --length LENGTH
                        percent overlap between target and query (default:
                        0.7)
  -v, --verbose         print more (default: False)
  -s, --separate-files  Output a separate file for each of the hits to a HMM
                        model. (default: False)
  -o OUT_FORMAT, --output-format OUT_FORMAT
                        The output format passed to hmmalign (default:
                        PSIBLAST)
  -c, --include-consensus
                        Include the HMM consensus sequence (generated by
                        hmmemit) in the alignment (default: False)
  -b, --best-hit        Only align the best match to the HMM even if there are
                        multiple matches that pass the threshold (default:
                        False)
</code></pre>

<h2>makeDB</h2>

<pre><code>checkM makeDB [-h] [-f] [-m] [-i IMG_BASE] [-M METADATA]
                     [--minimum-genome-count MIN_GENOME]
                     [--minimum-taxon-conservation MIN_CONS] [-p PFAM]
                     [-o DATABASE]

optional arguments:
  -h, --help            show this help message and exit
  -f, --force           overwite markers or taxons ifthey are already stored
                        in the database (default: False)
  -m, --make_tables     make new tables from scratch (default: False)
  -i IMG_BASE, --input-prefix IMG_BASE
                        The root directory to all the IMG genomes (default:
                        None)
  -M METADATA, --metadata METADATA
                        The IMG metadata file (default: None)
  --minimum-genome-count MIN_GENOME
                        A taxonomic lineage must have at least this number of
                        genomes for a marker set to be made (default: 5)
  --minimum-taxon-conservation MIN_CONS
                        Percent of genomes in taxon that must contain the
                        marker (default: 0.95)
  -p PFAM, --pfam-database PFAM
                        path to Pfam-A file (default: None)
  -o DATABASE, --output-database DATABASE
                        output sqlite3 database name (default: None)
</code></pre>

<h1>Licence and referencing</h1>

<p>Project home page, info on the source tree, documentation, issues and
how to contribute, see <a href="http://github.com/Ecogenomics/CheckM">http://github.com/Ecogenomics/CheckM</a></p>

<p>This software is currently unpublished. Please contact me at
m_dot_imelfort_at_uq_dot_edu_dot_au for more information about
referencing this software.</p>

<p>Copyright Â© 2012, 2013 Michael Imelfort, Connor Skennerton. See LICENSE.txt
for further details.</p>
      </section>
      <footer>
        <p>This project is maintained by <a href="https://github.com/Ecogenomics">Ecogenomics</a></p>
        <p><small>Hosted on GitHub Pages &mdash; Theme by <a href="https://github.com/orderedlist">orderedlist</a></small></p>
      </footer>
    </div>
    <script src="javascripts/scale.fix.js"></script>
    
  </body>
</html>