<!DOCTYPE html>
<html lang="en"><head><meta charset="utf-8"><title>CheckM - assessing the quality of genome bins</title><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="description" content=""><meta name="author" content=""><link href="./css/bootstrap.min.css" rel="stylesheet"><link href="./css/bootstrap-responsive.min.css" rel="stylesheet"><link href="./css/docs.css" rel="stylesheet"><!-- Le HTML5 shim, for IE6-8 support of HTML elements --><!--[if lt IE 9]>
<script src="http://html5shim.googlecode.com/svn/trunk/html5.js"></script>
<![endif]--><!-- Le styles --></head>
<body class="preview" data-spy="scroll" data-target=".subnav" data-offset="80">
<script src="./js/bsa.js"></script>
<div class="container">
<br>
<header class="jumbotron subhead" id="overview"> </header>
<div class="row"> <img style="display: block; margin-left: auto; margin-right: auto;" src="./img/checkm.png"> </div>
<section id="theboss"><!-- Headings & Paragraph Copy -->
</section>
<div class="row">
<div class="span4">
<hr>
<h1>Overview</h1>
<p>CheckM provides a set of tools for assessing the quality of genomes recovered from isolates,
single cells, or metagenomes. It provides robust estimates of genome completeness and contamination
by using collocated sets of genes that are ubiquitous and single-copy within a phylogenetic lineage.
Assessment of genome quality can also be examined using plots depicting key genomic characteristics
(e.g., GC, coding density) which highlight sequences outside the expected distributions of a typical
genome. CheckM also provides tools for identifying genome bins that are likely candidates for merging
based on marker set compatibility, similarity in genomic characteristics, and proximity within a reference
genome tree.</p>
</div>
<div class="span4">
<hr>
<h1>News</h1>
<p>
<ul><li>CheckM v1.1.6 was released on April 9, 2022 and requires Python 3.</li></ul>
</p>


<h1>Use CheckM</h1>
<p>Before using CheckM you need a set of putative genomes. These may come from isolates, single cells, or metagenomic data.
Our companion tool <a href="http://minillinim.github.io/GroopM/">GroopM</a> can be used to recover genomes from metagenomic
data.</p><p>For information on using CheckM visit the <a href="https://github.com/Ecogenomics/CheckM/wiki">wiki</a>.</p><h1>Cite CheckM</h1>
<p>If you find this software useful, we'd love for you to cite us:</p>
<ul><li>Parks DH, Imelfort M, Skennerton CT, Hugenholtz P, Tyson GW. 2014. <a href="http://genome.cshlp.org/content/25/7/1043">Assessing the quality of microbial genomes recovered
from isolates, single cells, and metagenomes.</a> Genome Research, 25: 1043-1055.</li></ul></div>
<div class="span4">
<hr><h1>Talk to us</h1>
<p>We'd love to hear from you. All comments and suggestions can be sent to Donovan Parks:<br>
</p>
<ul><li>donovan_dot_parks_at_gmail_dot_com</li></ul><h1>Licensing</h1>
<p>CheckM is licensed using the GNU General Public License
version 3 as published by the Free Software Foundation.</p>
<p>The CheckM logo is a product of Mike Imelfort's mind.</p>
<p>This site was created using a template created by the
wonderful people at <a href="http://bootswatch.com">bootswatch</a>.
</p>
</div>
<hr><!-- Placed at the end of the document so the pages load faster -->
<script src="http://ajax.googleapis.com/ajax/libs/jquery/1.7.2/jquery.min.js"></script>
<script src="./js/bootstrap.min.js"></script> </div>
</div></body></html>
